### PR TITLE
Add time io

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**
- Adds the io for the time_axis
- properly close the file
- bug fixes for the axis data/metadata writes
- set up time related variables in the file object
- Moves get_date_diff from diag_utils to diag_time_utils

Fixes # (issue)

**How Has This Been Tested?**
CI, including updating tests

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: diag_manager/fms_diag_file_object.F90 [hunk 1]
```diff
@@ -788,6 +824,132 @@ subroutine open_diag_file(this, time_step, file_is_opened)
...
+  character(len=50)                 :: time_units_str !< Time units written as a string
+  character(len=50)                 :: calendar       !< The calendar name
+
+  character(len=:), allocatable :: time_var_name !< The name of the time variable as it is defined in the yaml
+
+  diag_file => this%FMS_diag_file
+  fileobj => diag_file%fileobj
+
+  time_var_name = diag_file%get_file_unlimdim()
+  call register_axis(fileobj, time_var_name, unlimited)
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:842]: I may be missing this, but where is the variable `unlimited` coming from?

File: diag_manager/fms_diag_file_object.F90 [hunk 2]
```diff
@@ -788,6 +824,132 @@ subroutine open_diag_file(this, time_step, file_is_opened)
...
+!> \brief Determine if it is time to "write" to the file
+logical function is_time_to_write(this, time_step)
+  class(fmsDiagFileContainer_type), intent(in), target   :: this            !< The file object
+  TYPE(time_type),                  intent(in)           :: time_step       !< Current model step time
+
+  if (time_step >= this%FMS_diag_file%next_output) then
+    is_time_to_write = .true.
+    if (this%FMS_diag_file%is_static) return
+    if (time_step >= this%FMS_diag_file%next_next_output) &
+      call mpp_error(FATAL, this%FMS_diag_file%get_file_fname()//": You skip a time!")
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:879]: This error message made me lol, but it is not descriptive enough.  You should at the very least identify it as a diag manager error.

File: diag_manager/fms_diag_file_object.F90 (hunk 1 above)
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:842]: I may be missing this, but where is the variable `unlimited` coming from?

File: diag_manager/fms_diag_file_object.F90 [hunk 3]
```diff
@@ -788,6 +824,132 @@ subroutine open_diag_file(this, time_step, file_is_opened)
...
+  else
+    is_time_to_write = .false.
+  endif
+end function is_time_to_write
+
+!> \brief Determine if the current PE has data to write
+logical function writing_on_this_pe(this)
+  class(fmsDiagFileContainer_type), intent(in), target   :: this            !< The file object
+
+  writing_on_this_pe = .true.
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:889]: this should be moved into the select type under `default class`.
```fortran
select type(diag_file => this%FMS_diag_file)
   type is (subRegionalFile_type)
    writing_on_this_pe = diag_file%write_on_this_pe
   default class
     writing_on_this_pe = .true.
end select
```

File: diag_manager/fms_diag_file_object.F90 [hunk 4]
```diff
@@ -838,9 +1000,24 @@ subroutine write_axis_data(this, diag_axis)
...
-  call close_file(fileobj)
 end subroutine write_axis_data
 
+!< @brief Closes the diag_file
+subroutine close_diag_file(this)
+  class(fmsDiagFileContainer_type), intent(inout), target :: this            !< The file object
+
+  !< The select types are needed here because otherwise the code will go to the
+  !! wrong close_file routine and things will not close propertly
+  select type( fileobj => this%FMS_diag_file%fileobj)
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:1011]: You can't just call `close_file` here?  Shouldn't `close_file` have the `select type`?

File: diag_manager/fms_diag_time_utils.F90 [hunk 5]
```diff
@@ -27,17 +27,18 @@
 module fms_diag_time_utils_mod
 
 use time_manager_mod, only: time_type, increment_date, increment_time, get_calendar_type, NO_CALENDAR, leap_year, &
-                            get_date, get_time
+                            get_date, get_time,  operator(>), operator(<), operator(-)
 use diag_data_mod,    only: END_OF_RUN, EVERY_TIME, DIAG_SECONDS, DIAG_MINUTES, DIAG_HOURS, DIAG_DAYS, DIAG_MONTHS, &
                             DIAG_YEARS
 USE constants_mod,    ONLY: SECONDS_PER_DAY, SECONDS_PER_HOUR, SECONDS_PER_MINUTE
-use fms_mod,          only: fms_error_handler
+use fms_mod,          only: fms_error_handler, error_mesg, FATAL
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_time_utils.F90:34]: I understand why, but this is inconsistent with the other modules that use mpp_error

**Review by thomas-robinson — CHANGES_REQUESTED** [review]

File: diag_manager/fms_diag_file_object.F90 (hunk 1 above)
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:842]: `unlimited` comes fro fms2_io_mod

File: diag_manager/fms_diag_file_object.F90 (hunk 4 above)
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:1011]: `close_file` doesn't have the `select_type`, it has interface
https://github.com/NOAA-GFDL/FMS/blob/d8208e61be2dff46cc387cbace0f1f7c2bd950fc/fms2_io/fms2_io.F90#L166-L170

If I just call close_file, the code was always going into `netcdf_file_close_wrap`, so if the fileobject was `FmsNetcdfDomainFile_t`, some stuff doesn't get deallocated 
https://github.com/NOAA-GFDL/FMS/blob/d8208e61be2dff46cc387cbace0f1f7c2bd950fc/fms2_io/fms_netcdf_domain_io.F90#L60-L66

**Review by thomas-robinson — APPROVED** [review]

**Pull request merged**